### PR TITLE
[flutter_local_notifications_windows] Fix URIs on Windows -- images and audio

### LINKED ISSUE #2510: [Windows] How to play a custom sound?
**Describe the bug**
Hello, I’m trying to play a custom sound on Windows, but nothing I’ve tried is working. The README doesn’t explain how this should be achieved, and the example project `_showNotificationCustomSound()` only plays a preset sound: `WindowsNotificationAudio.preset(sound: WindowsNotificationSound.alarm5)`.

In the source code, I noticed a function that allows playing a sound from a URI: `WindowsNotificationAudio.fromFile`. However, I’m unsure of the exact path format required or how to test it. Additionally, in the `allowedSchemes`, I see that only `ms-appx and ms-resource` are supported. Does this mean I need to package it as an MSIX?

Any documentation or sample code would be great :)

**To Reproduce**

- Create a new notification with an audio file: 
```
WindowsNotificationDetails(
        audio: WindowsNotificationAudio.fromFile(file: Uri.parse("ms-appx://data/flutter_assets/assets/custom_sound.mp3"))
      )
```
- Error in the console:
```
flutter: Error :  Unsupported operation: Cannot extract a file path from a ms-appx URI
flutter: StackTrace :  #0      _SimpleUri.toFilePath (dart:core/uri.dart:4905:7)
#1      new WindowsNotificationAudio.fromFile (package:flutter_local_notifications_windows/src/details/notification_audio.dart:109:23)
#2      LocalNotificationsManager.schedule (package:projects/data/notifications/local_notifications_manager.dart:103:41)
#3      AlarmsManager.schedulePrayer (package:projects/data/alarms/alarms_manager.dart:118:28)
#4      AlarmsManager.showTestPrayer (package:projects/data/alarms/alarms_manager.dart:69:5)
```

**Expected behavior**
The custom sound should be played

### PR DESCRIPTION
**Pull request by Levi-Lesches**
Windows custom audio and image support wasn't very well tested, and #2510 reported that it doesn't even work in its current state. 

This PR: 
- removes all old image/audio constructors
- Adds `WindowsNotificationAudio.asset`, for MSIX apps only
- Adds `WindowsImage()`, for all sorts of URIs, with a big page of documentation
- Updates examples and tests accordingly
- Bumps versions and update changelog
- Adds a lot of doc comments but doesn't touch the README
  - See #2477 

Fixes #2510

### CODE REVIEW

File: flutter_local_notifications/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## [19.0.0-dev.4]
+
+* [Windows] Reworked the APIs around custom images and audio. Use the new `.asset()` or `.network()` constructors
```
Comment (MaikuB) [inline flutter_local_notifications/CHANGELOG.md:3]: Looks like the PR introduces breaking changes between prerelease versions. Can you update the changelog entries to call these out more explicitly?

**Comment by Levi-Lesches** [comment]
Umm... not sure why that's a "failure": 

```console
$ melos exec
  └> flutter test
     └> RUNNING (in 1 packages)

--------------------------------------------------------------------------------
flutter_local_notifications_windows:

✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: No details
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Simple details
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Actions
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Audio
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Rows
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Header
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Images
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Inputs
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Progress
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/bindings_test.dart: Bindings work in simple cases
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/bindings_test.dart: Bindings fail when ID is not found in simple cases
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/bindings_test.dart: Bindings are included in show()
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/bindings_test.dart: Bindings fail when notification has been cancelled
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin initializes safely
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin catches bad GUIDs
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin cannot be used before initializing
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin cannot be used after disposed
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin does not support repeating notifications
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/scheduled_test.dart: Schedules work with basic times
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/scheduled_test.dart: Schedules do not work with earlier time
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/xml_test.dart: XML catches invalid XML

🎉 [12](https://github.com/MaikuB/flutter_local_notifications/actions/runs/12556229602/job/35007235468?pr=2511#step:6:13) tests passed.
--------------------------------------------------------------------------------

$ melos exec
  └> flutter test
     └> FAILED (in 1 packages)
        └> flutter_local_notifications_windows (with exit code 1)

melos run test:unit:windows
  └> melos exec -c 1 -- "flutter test"
     └> FAILED
ScriptException: The script test:unit:windows failed to execute.
Error: Process completed with exit code 1.
```

File: flutter_local_notifications/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+## [19.0.0-dev.4]
+
+* [Windows] Reworked the APIs around custom images and audio. Use the new `.asset()` or `.network()` constructors. This is a breaking change from `19.0.0-dev.1`.
```
Comment (MaikuB) [inline flutter_local_notifications/CHANGELOG.md:3]: Looks like there's breaking changes between different prereleases versions. Can you update the changelog entry to be more explicit in calling these out?

Comment (Levi-Lesches) [reply flutter_local_notifications/CHANGELOG.md:3]: Oh, sorry, did this in the main package and not the sub-package. Fixed!

Comment (Levi-Lesches) [reply flutter_local_notifications/CHANGELOG.md:3]: Oh wait did you want me to elaborate more on the top-level package as well?

Comment (MaikuB) [reply flutter_local_notifications/CHANGELOG.md:3]: It should apply to both. I was referring to a couple of things though on this though

- breaking changes are indicated via `**Breaking changes**` for multiple things or `**Breaking change**` when there's a single breaking change. There are existing examples in other changelog entries
- details on what's been removed
- a note that the example app being updated to demonstrate usages could be worth mentioning too

File: flutter_local_notifications_windows/lib/src/details/notification_audio.dart [hunk 3]
```diff
@@ -101,41 +98,22 @@ class WindowsNotificationAudio {
...
+  ///
+  /// Note that this will only work in release builds that have been packaged as
+  /// an MSIX installer. Currently, there is no way to test this in debug mode.
+  ///
+  /// Windows supports the following formats: `.aac`, `.flac`, `.m4a`, `.mp3`,
+  /// `.wav`, and `.wma`.
+  WindowsNotificationAudio.asset(
+    String assetName, {
     this.shouldLoop = false,
+    WindowsNotificationSound debugModeFallback =
```
Comment (MaikuB) [inline flutter_local_notifications_windows/lib/src/details/notification_audio.dart:111]: Since `debugModeFallback` isn't a property, can you update the constructor's API docs to explain what `debugModeFallback` is used for? I believe that's the only place to do so as I tried using `///` above the line and it wasn't included in the generated docs. If it helps copilot came up with `The [debugModeFallback] sound will be used in debug mode.` that can be added and this was the result

image

File: flutter_local_notifications/CHANGELOG.md (hunk 2 above)
Comment (Levi-Lesches) [reply flutter_local_notifications/CHANGELOG.md]: Ah okay. I figured since it was just a breakage on a _pre-release_ it was okay, but I added that stuff now.

File: flutter_local_notifications_windows/lib/src/details/notification_audio.dart (hunk 3 above)
Comment (Levi-Lesches) [reply flutter_local_notifications_windows/lib/src/details/notification_audio.dart:113]: (Done)

**Comment by MaikuB** [comment]
I'm not sure why it's failing as well but seems like some tests are failing to  completely run

**Comment by MaikuB** [comment]
Actually from the output you've shared from running it locally, it still has an exit code of 1 so that would be why. Not sure why that is happening but if that's fixed locally then that would most likely fix the issue

**Comment by Levi-Lesches** [comment]
I edited my message after but the console output is just from GitHub.

**Comment by Levi-Lesches** [comment]
I went back and re-worked things slightly based on your action-image comment: both `WindowsImage` and `WindowsAction` now support any `Uri` in the constructor, with an extensive docs page on when to use which kind of URI, and a function `WindowsImage.assetUri()` that gives you a safe URI of any Flutter asset. 

I also updated my previous message with my actual (successful) test output on my own laptop

I'm also open to renaming everything `WindowsNotificationSomething` to be consistent. It might be good to get that out before the pre-releases go stable.

**Comment by MaikuB** [comment]
I'm going to push a change to temporarily enable verbose logging on the tests to see if it might provide more info on what's going on

**Comment by MaikuB** [comment]
Actually I won't need to do that anymore. I've pasted the output below

```
PS C:\flutter_local_notifications\flutter_local_notifications_windows> flutter test
00:06 +19 -1: C:/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Progress [E]
  Expected: NotificationUpdateResult:<NotificationUpdateResult.success>
    Actual: NotificationUpdateResult:<NotificationUpdateResult.notFound>

  package:matcher                expect
  test\details_test.dart 243:11  main.<fn>.<fn>

00:06 +19 -1: C:/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Progress
Retry: Details: Progress
00:07 +19 -1: C:/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Progress [E]
  Expected: NotificationUpdateResult:<NotificationUpdateResult.success>
    Actual: NotificationUpdateResult:<NotificationUpdateResult.notFound>

  package:matcher                expect
  test\details_test.dart 243:11  main.<fn>.<fn>

Retry: Details: Progress
00:08 +20 -1: C:/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart: Details: Progress [E]
  Expected: NotificationUpdateResult:<NotificationUpdateResult.success>
    Actual: NotificationUpdateResult:<NotificationUpdateResult.notFound>

  package:matcher                expect
  test\details_test.dart 243:11  main.<fn>.<fn>

Retry: Details: Progress
00:08 +21: All tests passed!
```

The summary says all tests pass but it actually picked up differences between the expected and actual values. I could be wrong as it's been a while but I don't recall this being an issue before. Perhaps this is the reason why GitHub found issues?

**Comment by Levi-Lesches** [comment]
Well, I didn't get anything interesting by checking out the lines the stack trace is referring to. The tests pass, no errors occur, but the exit code is being set somewhere to a non-zero value. But it's using the global variable so it's a little hard to follow the logic.

**Comment by MaikuB** [comment]
I also took a look at the verbose logs and looks like the tooling is crapping out. Not sure why but may have to ignore this for now. As for verifying the changes in the PR, I've tried to create the MSIX for the example but can't install it. I changed the MSIX config to install the cert too but that didn't help. Any pointers?

**Comment by Levi-Lesches** [comment]
If you're getting the "untrusted certificate" error, or something like that when trying to install a valid msix, that's normal:

- Right click on the msix, click Properties
- digital signatures tab, choose Levi Lesches, click Details
- click View Certificate, then Install Certificate
- Install the certificate in Local Machine, then click Browse and choose Trusted Root

It seems to be another deterrent to stop people from running unsigned code. Users who actually have a valid and not just self-signed MSIX certificate won't run into these issues

**Comment by Levi-Lesches** [comment]
Okay I submitted a few PRs to `package:msix`:

- https://github.com/YehudaKremer/msix/pull/293
- https://github.com/YehudaKremer/msix/pull/294

Using these functions, we can provide a better experience for `WindowsImage.assetUri` that provides the correct path on debug builds, release builds, and release builds in an MSIX. It uses new functions from `package:win32`, which requires dart: ^3.4.0` though

**Comment by MaikuB** [comment]
Did I miss something as the custom sound was playing for me when I ran a debug build via F5 and VS code prior to the recent commit?

If these functions are needed to provide a better experience and solve the problem, can the functionality be added to the plugin directly without depending on the `msix` package? This helps minimise the dependencies that ideally should be confined to first-party ones, and reduce the chance for version conflicts. Given, `msix`  itself also depends on a lot of other packages, the latter will more likely happen e.g. for those that use `get_it` in the app already

**Comment by Levi-Lesches** [comment]
> Did I miss something as the custom sound was playing for me when I ran a debug build via F5 and VS code prior to the recent commit?

I was actually referring to images. For an asset `assets/image.png`,
- in debug mode, use `file:///<project>/assets/image.png`
- in release mode (exe), use `file:///<exe>/data/flutter_assets/assets/image.png`
- in release mode (msix), use `ms-appx:///data/flutter_assets/image.png`

After extensive testing, this is really the only combo that reliably works in every situation. Users can use `WindowsImage.assetUri()` to get the right URI automatically. 

For audio, Windows only supports custom files in `ms-appx://` format, so there is no such debug/exe support, and we have to use a preset in those cases. I made audio use the fallback for release builds (exe) as well. 

> If these functions are needed to provide a better experience and solve the problem, can the functionality be added to the plugin directly without depending on the `msix` package?

Done -- added a new class, `MsixUtils`, with a conditional import on `dart:ffi`. I wanted to try to add it to the Windows plugin so that people can use it with the `plugin?.resolve<Windows>()` style, but since `WindowsImage.assetUri()` and `WindowsNotificationAudio.asset()` use these at the top-level, they had to be static. I also disabled the `avoid_classes_with_only_static_members` lint because the functions are inherently static, and it's useful to keep them under a namespace.

**Comment by Levi-Lesches** [comment]
Found the test causing the tooling to freak out. 

```dart
// test/scheduled_test.dart
test('work with basic times', skip: true, () async {
  await plugin.cancelAll();
  expect(await countPending(), 0);
  final TZDateTime now = TZDateTime.now(location);
  final TZDateTime later = now.add(const Duration(days: 1));
  expect(plugin.zonedSchedule(300, null, null, later, null), completes);
  expect(await countPending(), 1);
  expect(plugin.zonedSchedule(301, null, null, later, null), completes);
  expect(await countPending(), 2);
  expect(plugin.zonedSchedule(302, null, null, later, null), completes);
  expect(await countPending(), 3);
  await plugin.cancelAll();
  expect(await countPending(), 0);
});
```

Here's my local output: 

```console
[   +6 ms] test 3: Started flutter_tester process at pid 20056
[ +180 ms] test 3: connected to test device, now awaiting test result
[   +1 ms] test 3: Waiting for test harness or tests to finish
✅ D:/Flutter/Packages/flutter_local_notifications/flutter_local_notifications_windows/test/scheduled_test.dart: Schedules work with basic times
✅ D:/Flutter/Packages/flutter_local_notifications/flutter_local_notifications_windows/test/scheduled_test.dart: Schedules do not work with earlier time
[ +191 ms] test 3: Test harness is no longer needed by test process
[        ] test 3: finished
[   +1 ms] test 3: cleaning up...
[        ] test 3: ensuring test device is terminated.
[        ] test 3: Terminating flutter_tester process
[        ] test 3: Shutting down DevTools server
[        ] test 3: Test process is no longer needed by test harness
[        ] test 3: Shutting down test harness socket server
[   +6 ms] test 3: flutter_tester process at pid 20056 exited with code=-1
[        ] test 3: deleting temporary directory
[   +2 ms] test 3: finished
```

Here's the CI output: 

```console
[   +2 ms] test 3: Started flutter_tester process at pid 4008
[ +189 ms] test 3: connected to test device, now awaiting test result
[        ] test 3: Waiting for test harness or tests to finish
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/scheduled_test.dart: Schedules work with basic times
[+1226 ms] test 3: Test process is no longer needed by test harness
[        ] test 3: finished
[        ] test 3: cleaning up...
[        ] test 3: ensuring test device is terminated.
[        ] test 3: Terminating flutter_tester process
[        ] test 3: Shutting down DevTools server
[        ] test 3: Test harness is no longer needed by test process
✅ d:/a/flutter_local_notifications/flutter_local_notifications/flutter_local_notifications_windows/test/scheduled_test.dart: Schedules do not work with earlier time
[        ] test 3: Shutting down test harness socket server
[        ] Compiling file:///C:/Users/RUNNER~1/AppData/Local/Temp/flutter_tools.bce9bb2d/flutter_test_listener.20e0f0d4/listener.dart
[   +3 ms] test 3: flutter_tester process at pid 4008 exited with code=-1073740791
[        ] test 3: deleting temporary directory
[   +1 ms] test 3: finished
```

Locally, the tests run together. On CI, the tests are interrupted and the process is shut down early. Checking the Dart and C++ code, and cross-referencing that with what else works on CI, my guess is that one of these two lines is causing the crash: 

```cpp
  ScheduledToastNotification notification(doc, winrt::clock::from_time_t(time));
  plugin->notifier.value().AddToSchedule(notification);
```

And somehow the Flutter tooling isn't picking it up as a failed test. Then, when it reaches the end and notices that a test didn't finish properly somewhere, it exits with a failure. I'll try to report this to the Flutter team soon, but for now I added `skip: true` to that test.

File: flutter_local_notifications_windows/lib/src/details/notification_parts.dart [hunk 4]
```diff
@@ -29,24 +32,67 @@ enum WindowsImageCrop {
...
     this.placement,
     this.crop,
   });
 
+  /// Creates a URI for a [Flutter asset](https://docs.flutter.dev/ui/assets/assets-and-images#loading-images).
+  ///
+  /// - In debug mode, resolves to a file URI to the asset itself
+  /// - In non-MSIX release builds, resolves to a file URI to the bundled asset
+  /// - In MSIX releases, resolves to an `ms-appx` URI from [Msix.assetUri].
+  static Uri assetUri(String assetName) {
```
Comment (MaikuB) [inline flutter_local_notifications_windows/lib/src/details/notification_parts.dart:75]: As this isn't a constructor but a method, can you rename it so it reads like one to convey that an action is being taken? I was originally thinking `resolve` should be part of it but saw [this](https://api.flutter.dev/flutter/dart-ui_web/AssetManager/getAssetUrl.html) exists that `getAssetUri()` could potentially work too. If you use the latter though then would be appropriate to update the API docs e.g. replace "creates" with "returns".

Similar comment applies to the method in `MsixUtils`

File: flutter_local_notifications_windows/lib/src/plugin/stub.dart [hunk 5]
```diff
@@ -96,4 +96,7 @@ class FlutterLocalNotificationsWindows extends WindowsNotificationsBase {
   @override
   @visibleForTesting
   void enableMultithreading() {}
```
Comment (MaikuB) [inline flutter_local_notifications_windows/lib/src/plugin/stub.dart:98]: Based on your recent findings, is this method still needed?

File: flutter_local_notifications_windows/dart_test.yaml [hunk 6]
```diff
@@ -1,3 +1,3 @@
 platforms: [vm]
 test_on: windows
-retry: 5  # These tests have concurrency issues. See bin/crash.dart
+# retry: 5  # These tests have concurrency issues. See bin/crash.dart
```
Comment (MaikuB) [inline flutter_local_notifications_windows/dart_test.yaml:3]: As there's source control, can you remove this over commenting it out?

File: flutter_local_notifications_windows/lib/src/plugin/stub.dart (hunk 5 above)
Comment (Levi-Lesches) [reply flutter_local_notifications_windows/lib/src/plugin/stub.dart:98]: Not _needed_, as we can use `-j 1`, but I don't think that solves the fundamental problem of concurrency at hand. I can remove it though as it really shouldn't be needed -- or feel free to remove it yourself if you'd like.

Comment (Levi-Lesches) [reply flutter_local_notifications_windows/lib/src/plugin/stub.dart]: Removed

File: flutter_local_notifications_windows/lib/src/details/notification_parts.dart (hunk 4 above)
Comment (Levi-Lesches) [reply flutter_local_notifications_windows/lib/src/details/notification_parts.dart]: Done

**Comment by MaikuB** [comment]
Whilst PR is green I'm getting the following output from running the tests locally

```
melos run test:unit:windows
  └> melos exec -c 1 -- "flutter test -j 1"
     └> RUNNING

$ melos exec
  └> flutter test -j 1
     └> RUNNING (in 1 packages)

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
flutter_local_notifications_windows:
00:00 +0: loading c:/flutter_local_notifications/flutter_local_notifications_windows/test/bindings_test.dart
00:00 +0 -1: loading c:/flutter_local_notifications/flutter_local_notifications_windows/test/bindings_test.dart [E]
  Failed to load "c:/flutter_local_notifications/flutter_local_notifications_windows/test/bindings_test.dart":
  Invalid argument(s): Failed to load dynamic library 'flutter_local_notifications_windows.dll': %1 is not a valid Win32 application.
   (error code: 193)
  dart:ffi                                                               new DynamicLibrary.open
  package:flutter_local_notifications_windows/src/plugin/ffi.dart 44:22  new FlutterLocalNotificationsWindows
  test\bindings_test.dart 18:11                                          main.<fn>
  package:test_api                                                       Declarer.group
  package:test_core/src/scaffolding.dart 224:13                          group
  test\bindings_test.dart 15:16                                          main

00:00 +0 -1: loading c:/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart
00:00 +0 -2: loading c:/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart [E]
  Failed to load "c:/flutter_local_notifications/flutter_local_notifications_windows/test/details_test.dart":
  Invalid argument(s): Failed to load dynamic library 'flutter_local_notifications_windows.dll': %1 is not a valid Win32 application.
   (error code: 193)
  dart:ffi                                                               new DynamicLibrary.open
  package:flutter_local_notifications_windows/src/plugin/ffi.dart 44:22  new FlutterLocalNotificationsWindows
  test\details_test.dart 30:11                                           main.<fn>
  package:test_api                                                       Declarer.group
  package:test_core/src/scaffolding.dart 224:13                          group
  test\details_test.dart 27:16                                           main

00:00 +0 -2: loading c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart
00:00 +0 -2: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin (setUpAll)
00:00 +0 -2: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin initializes safely
00:00 +0 -3: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin initializes safely [E]
  Invalid argument(s): Failed to load dynamic library 'flutter_local_notifications_windows.dll': %1 is not a valid Win32 application.
   (error code: 193)
  dart:ffi                                                               new DynamicLibrary.open
  package:flutter_local_notifications_windows/src/plugin/ffi.dart 44:22  new FlutterLocalNotificationsWindows
  test\plugin_test.dart 27:13                                            main.<fn>.<fn>

00:00 +0 -3: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin catches bad GUIDs
00:00 +0 -4: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin catches bad GUIDs [E]
  Invalid argument(s): Failed to load dynamic library 'flutter_local_notifications_windows.dll': %1 is not a valid Win32 application.
   (error code: 193)
  dart:ffi                                                               new DynamicLibrary.open
  package:flutter_local_notifications_windows/src/plugin/ffi.dart 44:22  new FlutterLocalNotificationsWindows
  test\plugin_test.dart 35:13                                            main.<fn>.<fn>

00:00 +0 -4: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin cannot be used before initializing
00:00 +0 -5: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin cannot be used before initializing [E]
  Invalid argument(s): Failed to load dynamic library 'flutter_local_notifications_windows.dll': %1 is not a valid Win32 application.
   (error code: 193)
  dart:ffi                                                               new DynamicLibrary.open
  package:flutter_local_notifications_windows/src/plugin/ffi.dart 44:22  new FlutterLocalNotificationsWindows
  test\plugin_test.dart 42:13                                            main.<fn>.<fn>

00:00 +0 -5: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin cannot be used after disposed
00:00 +0 -6: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin cannot be used after disposed [E]
  Invalid argument(s): Failed to load dynamic library 'flutter_local_notifications_windows.dll': %1 is not a valid Win32 application.
   (error code: 193)
  dart:ffi                                                               new DynamicLibrary.open
  package:flutter_local_notifications_windows/src/plugin/ffi.dart 44:22  new FlutterLocalNotificationsWindows
  test\plugin_test.dart 70:13                                            main.<fn>.<fn>

00:00 +0 -6: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin does not support repeating notifications
00:00 +0 -7: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin does not support repeating notifications [E]
  Invalid argument(s): Failed to load dynamic library 'flutter_local_notifications_windows.dll': %1 is not a valid Win32 application.
   (error code: 193)
  dart:ffi                                                               new DynamicLibrary.open
  package:flutter_local_notifications_windows/src/plugin/ffi.dart 44:22  new FlutterLocalNotificationsWindows
  test\plugin_test.dart 98:13                                            main.<fn>.<fn>

00:00 +0 -7: c:/flutter_local_notifications/flutter_local_notifications_windows/test/plugin_test.dart: Plugin (tearDownAll)
00:00 +0 -7: loading c:/flutter_local_notifications/flutter_local_notifications_windows/test/scheduled_test.dart
00:01 +0 -8: loading c:/flutter_local_notifications/flutter_local_notifications_windows/test/scheduled_test.dart [E]
  Failed to load "c:/flutter_local_notifications/flutter_local_notifications_windows/test/scheduled_test.dart":
  Invalid argument(s): Failed to load dynamic library 'flutter_local_notifications_windows.dll': %1 is not a valid Win32 application.
   (error code: 193)
  dart:ffi                                                               new DynamicLibrary.open
  package:flutter_local_notifications_windows/src/plugin/ffi.dart 44:22  new FlutterLocalNotificationsWindows
  test\scheduled_test.dart 13:11                                         main.<fn>
  package:test_api                                                       Declarer.group
  package:test_core/src/scaffolding.dart 224:13                          group
  test\scheduled_test.dart 11:16                                         main

00:01 +0 -8: loading c:/flutter_local_notifications/flutter_local_notifications_windows/test/xml_test.dart
00:01 +0 -9: loading c:/flutter_local_notifications/flutter_local_notifications_windows/test/xml_test.dart [E]
  Failed to load "c:/flutter_local_notifications/flutter_local_notifications_windows/test/xml_test.dart":
  Invalid argument(s): Failed to load dynamic library 'flutter_local_notifications_windows.dll': %1 is not a valid Win32 application.
   (error code: 193)
  dart:ffi                                                               new DynamicLibrary.open
  package:flutter_local_notifications_windows/src/plugin/ffi.dart 44:22  new FlutterLocalNotificationsWindows
  test\xml_test.dart 66:9                                                main.<fn>
  package:test_api                                                       Declarer.group
  package:test_core/src/scaffolding.dart 224:13                          group
  test\xml_test.dart 62:3                                                main

00:01 +0 -9: Some tests failed.
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

$ melos exec
  └> flutter test -j 1
     └> FAILED (in 1 packages)
        └> flutter_local_notifications_windows (with exit code 1)

melos run test:unit:windows
  └> melos exec -c 1 -- "flutter test -j 1"
     └> FAILED
ScriptException: The script test:unit:windows failed to execute.
```

**Comment by Levi-Lesches** [comment]
Interesting, mine works fine: 

```console
$ flutter test -j 1
00:03 +20: All tests passed!
```

Google searches indicate you're trying to load a 32-bit DLL -- can you delete it and try running `build.bat` again? Try to see from the CMake output if it's 64 or 32 bit

**Comment by MaikuB** [comment]
This is the output

```
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The CXX compiler identification is MSVC 19.42.34433.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.42.34433/bin/Hostarm64/arm64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning (dev) at CMakeLists.txt:18 (set):
  Cannot set "flutter_local_notifications_windows_bundled_libraries": current
  scope has no parent.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring done (1.5s)
-- Generating done (0.0s)
-- Build files have been written to: C:/flutter_local_notifications/flutter_local_notifications_windows/build
MSBuild version 17.12.6+db5f6012c for .NET Framework

  1>Checking Build System
  Building Custom Rule C:/flutter_local_notifications/flutter_local_notifications_windows/src/CMakeLists.txt
  ffi_api.cpp
  plugin.cpp
  utils.cpp
  Generating Code...
     Creating library C:/flutter_local_notifications/flutter_local_notifications_windows/build/shared/Debug/flutter_local_notifications_windows.lib and object C:/flutter_local_notifica
  tions/flutter_local_notifications_windows/build/shared/Debug/flutter_local_notifications_windows.exp
  flutter_local_notifications_windows.vcxproj -> C:\flutter_local_notifications\flutter_local_notifications_windows\build\shared\Debug\flutter_local_notifications_windows.dll
  Building Custom Rule C:/flutter_local_notifications/flutter_local_notifications_windows/windows/CMakeLists.txt
        1 file(s) copied.
```

I'm assuming it's building from ARM64 (note: I'm running Windows 11 via Parallels on my M2 MacBook). I went back to earlier commits, even on master and still have the same issue so wonder if I've forgotten that there was a trick that I had to get it working before that I've forgotten about. I might have ignore this for now and figure out a different approach e.g. get a 32 bit DLL elsewhere.

On a different note, when doing a searching through your current fork, there's still some references to the `enableMultithreading()` method, most are ones that are commented out from the test that can be removed but others from actual code

image

**Comment by Levi-Lesches** [comment]
> I'm assuming it's building from ARM64 (note: I'm running Windows 11 via Parallels on my M2 MacBook).

Interesting. Not sure how you were doing that before but I can send you the 64-bit DLL if you need. 

> there's still some references to the enableMultithreading() method

Thanks 🤦, removed from the C++ side as well

File: flutter_local_notifications_windows/test/xml_test.dart [hunk 7]
```diff
@@ -56,25 +58,24 @@ const String complexXml = '''
...
-        expect(plugin.showRawXml(id: 1, xml: invalidXml), throwsArgumentError);
-        expect(
-            plugin.showRawXml(id: 2, xml: notWindowsXml), throwsArgumentError);
-        expect(
-            plugin.showRawXml(id: 3, xml: unmatchedXml), throwsArgumentError);
-        expect(plugin.showRawXml(id: 4, xml: validXml), completes);
-        expect(plugin.showRawXml(id: 5, xml: complexXml), completes);
-      });
+    test('catches invalid XML', () async {
+      expect(plugin.isValidXml(emptyXml), isFalse);
```
Comment (MaikuB) [inline flutter_local_notifications_windows/test/xml_test.dart:73]: Something I picked up after merging but these are all meant to be separate tests as there's a different permutations of input and output but have made as part of a single `test()`. The description given is also misleading as there are tests that assert valid XML was provided and is more appropriate as a group description. I noticed there are other test classes that are like this too

Comment (Levi-Lesches) [reply flutter_local_notifications_windows/test/xml_test.dart:73]: Yes I plan to re-organize the tests, and have more thorough ones. For example, while there are tests to make sure the APIs can distinguish between valid and invalid XML, and tests to make sure the `WindowsNotificationsDetails` generates valid XML, there are no tests to map `WindowsNotificationDetails` to specific XML, eg, to verify that a progress bar is actually inserted. Another PR for another day :)
